src/lib.rs

```diff
@@ -124,96 +24,1138 @@ impl<T> RpcMessage for T where
+        use crate::sealed::Sealed;
+
+        #[derive(Debug)]
+        pub struct NoSender;
+        impl Sealed for NoSender {}
+        impl crate::Sender for NoSender {}
+
+        #[derive(Debug)]
+        pub struct NoReceiver;
+
+        impl Sealed for NoReceiver {}
+        impl crate::Receiver for NoReceiver {}
+    }
+
+    /// Error when sending a oneshot or spsc message. For local communication,
+    /// the only thing that can go wrong is that the receiver has been dropped.
+    ///
+    /// For rpc communication, there can be any number of errors, so this is a
+    /// generic io error.
+    #[derive(Debug, thiserror::Error)]
+    pub enum SendError {
+        #[error("receiver closed")]
+        ReceiverClosed,
+        #[error("io error: {0}")]
+        Io(#[from] io::Error),
+    }
+
+    impl From<SendError> for io::Error {
+        fn from(e: SendError) -> Self {
+            match e {
+                SendError::ReceiverClosed => io::Error::new(io::ErrorKind::BrokenPipe, e),
+                SendError::Io(e) => e,
+            }
+        }
+    }
+
+    /// Error when receiving a oneshot or spsc message. For local communication,
+    /// the only thing that can go wrong is that the sender has been closed.
+    ///
+    /// For rpc communication, there can be any number of errors, so this is a
+    /// generic io error.
+    #[derive(Debug, thiserror::Error)]
+    pub enum RecvError {
+        #[error("sender closed")]
+        SenderClosed,
+        #[error("io error: {0}")]
+        Io(#[from] io::Error),
+    }
+
+    impl From<RecvError> for io::Error {
+        fn from(e: RecvError) -> Self {
+            match e {
+                RecvError::Io(e) => e,
+                RecvError::SenderClosed => io::Error::new(io::ErrorKind::BrokenPipe, e),
+            }
+        }
+    }
+}
+
+/// A wrapper for a message with channels to send and receive it.
+/// This expands the protocol message to a full message that includes the
+/// active and unserializable channels.
 ///
-/// Depending on the interaction type, you might need to implement traits that further
-/// define details of the interaction.
+/// rx and tx can be set to an appropriate channel kind.
+pub struct WithChannels<I: Channels<S>, S: Service> {
+    /// The inner message.
+    pub inner: I,
+    /// The return channel to send the response to. Can be set to [`crate::channel::none::NoSender`] if not needed.
+    pub tx: <I as Channels<S>>::Tx,
+    /// The request channel to receive the request from. Can be set to [`NoReceiver`] if not needed.
+    pub rx: <I as Channels<S>>::Rx,
+    /// The current span where the full message was created.
+    #[cfg(feature = "message_spans")]
+    #[cfg_attr(quicrpc_docsrs, doc(cfg(feature = "message_spans")))]
+    pub span: tracing::Span,
+}
+
+impl<I: Channels<S> + Debug, S: Service> Debug for WithChannels<I, S> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("")
```

Comment:
why a tuple, not a struct?

Reply:
You usually create these using a From conversion from a tuple, and having the debug output be a struct makes the debug output very long, with stuff like inner = ..., tx = ..., rx = ...(, span = ...).

Just a matter of taste I guess.